### PR TITLE
Optimize max_execution_date query in single dag case

### PR DESCRIPTION
**Pull request by joshowen**
Optimized queryplan to avoid single item groupby which isn't able to efficiently use the `dag_run_dag_id_execution_date_key` index:

<details>
<summary>Before:</summary>

| QUERY PLAN |
| --- |
| Nested Loop  \(cost=1000.55..367593.28 rows=17 width=189\) \(actual time=1862.277..1862.420 rows=1 loops=1\) |
|   \-\>  Finalize GroupAggregate  \(cost=1000.00..366374.21 rows=142 width=28\) \(actual time=1862.234..1862.376 rows=1 loops=1\) |
|         Group Key: dag\_run\_1.dag\_id |
|         \-\>  Gather  \(cost=1000.00..366371.37 rows=284 width=28\) \(actual time=1855.233..1862.361 rows=3 loops=1\) |
|               Workers Planned: 2 |
|               Workers Launched: 2 |
|               \-\>  Partial GroupAggregate  \(cost=0.00..365342.97 rows=142 width=28\) \(actual time=1849.562..1849.563 rows=1 loops=3\) |
|                     Group Key: dag\_run\_1.dag\_id |
|                     \-\>  Parallel Seq Scan on dag\_run dag\_run\_1  \(cost=0.00..363162.04 rows=435902 width=28\) \(actual time=271.546..1736.712 rows=351110 loops=3\) |
|                           Filter: \(\(\(dag\_id\)::text = 'REDACTED'::text\) AND \(\(\(run\_type\)::text = 'backfill'::text\) OR \(\(run\_type\)::text = 'scheduled'::text\)\)\) |
|                           Rows Removed by Filter: 782975 |
|   \-\>  Index Scan using dag\_run\_dag\_id\_execution\_date\_key on dag\_run  \(cost=0.56..8.57 rows=1 width=189\) \(actual time=0.036..0.036 rows=1 loops=1\) |
|         Index Cond: \(\(\(dag\_id\)::text = \(dag\_run\_1.dag\_id\)::text\) AND \(execution\_date = \(max\(dag\_run\_1.execution\_date\)\)\)\) |
| Planning Time: 0.369 ms |
| Execution Time: 1862.473 ms |

</details>

<details>
<summary>After:</summary>

| QUERY PLAN |
| --- |
| Nested Loop  \(cost=2.50..7.28 rows=5 width=189\) \(actual time=0.048..0.050 rows=1 loops=1\) |
|   \-\>  Result  \(cost=1.95..1.96 rows=1 width=8\) \(actual time=0.038..0.039 rows=1 loops=1\) |
|         InitPlan 1 \(returns $0\) |
|           \-\>  Limit  \(cost=0.56..1.95 rows=1 width=8\) \(actual time=0.037..0.037 rows=1 loops=1\) |
|                 \-\>  Index Scan Backward using dag\_run\_dag\_id\_execution\_date\_key on dag\_run dag\_run\_1  \(cost=0.56..1456402.98 rows=1046683 width=8\) \(actual time=0.036..0.036 rows=1 loops=1\) |
|                       Index Cond: \(\(\(dag\_id\)::text = 'REDACTED'::text\) AND \(execution\_date IS NOT NULL\)\) |
|                       Filter: \(\(\(run\_type\)::text = 'backfill'::text\) OR \(\(run\_type\)::text = 'scheduled'::text\)\) |
|   \-\>  Index Scan using dag\_run\_dag\_id\_execution\_date\_key on dag\_run  \(cost=0.56..24.37 rows=5 width=189\) \(actual time=0.008..0.008 rows=1 loops=1\) |
|         Index Cond: \(\(\(dag\_id\)::text = 'REDACTED'::text\) AND \(execution\_date = \($0\)\)\) |
| Planning Time: 0.403 ms |
| Execution Time: 0.087 ms |

</details>

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by uranusjr** [comment]
Instead o duplicating much of the query in two branches, I’d extract the actually different part (there `where` argument) instead.

**Comment by uranusjr** [comment]
Also please set up pre-commit according to the contribution guide and fix linting issues.

**Comment by joshowen** [comment]
@uranusjr I didn't see a clean way to avid creating 2 branches of the query while maintaining readability.  This change touches both the subquery and the where clause of the parent.

File: airflow/models/dag.py [hunk 1]
```diff
@@ -2925,21 +2925,44 @@ def bulk_write_to_db(
...
-                .group_by(DagRun.dag_id)
-                .subquery()
-            )
-            most_recent_runs_iter = session.scalars(
-                select(DagRun).where(
-                    DagRun.dag_id == most_recent_subq.c.dag_id,
-                    DagRun.execution_date == most_recent_subq.c.max_execution_date,
+                most_recent_runs_iter = session.scalars(
+                    select(DagRun).where(
+                        DagRun.dag_id == existing_dags[0],
```
Comment (uranusjr) [inline airflow/models/dag.py]: Would the optimization work if we use `DagRun.dag_id == most_recent_subq.c.dag_id` for both cases? If so we can reduce the code a lot.

Comment (joshowen) [reply airflow/models/dag.py]: That would still require the Group By in the subquery, which is what the optimized path avoids.

Comment (uranusjr) [reply airflow/models/dag.py]: So the answer is no?

Comment (joshowen) [reply airflow/models/dag.py]: I think the answer is no.

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
I think we need a test case to cover the one DAG case.

**Review by amoghrajesh — COMMENTED** [review]
Hello, I think it would be nice to add a specific test case for this. Wdyt @uranusjr?

**Comment by uranusjr** [comment]
Both branches should be covered by tests, whichever covers which.

**Comment by joshowen** [comment]
@uranusjr @amoghrajesh @eladkal I've added tests to cover both paths and rebased.

**Comment by dstandish** [comment]
This is a good optimization.  We ran into this exact issue and came up with the same optimization independently.  But we were thinking to just add an index since there are a couple other instances of if very similar query that I don't think can do the same thing.

But even if we do add the index I don't think it hurts to have both.

My one suggestion would be to move the query generation to a private function somewhere so that things are more readable at the call site.

File: airflow/models/dag.py [hunk 2]
```diff
@@ -3068,21 +3068,45 @@ def bulk_write_to_db(
...
         if any(dag.timetable.can_be_scheduled for dag in dags):
             # Get the latest automated dag run for each existing dag as a single query (avoid n+1 query)
-            last_automated_runs_subq = (
-                select(DagRun.dag_id, func.max(DagRun.execution_date).label("max_execution_date"))
-                .where(
-                    DagRun.dag_id.in_(existing_dags),
-                    or_(DagRun.run_type == DagRunType.BACKFILL_JOB, DagRun.run_type == DagRunType.SCHEDULED),
+            if len(existing_dags) == 1:
+                # Index optimized fast path to avoid more complicated & slower groupby queryplan
+                existing_dag_id = list(existing_dags)[0]
```
Comment (dstandish) [inline airflow/models/dag.py]:
```suggestion
                existing_dag_id = list(existing_dags)[0].dag_id
```

**Review by dstandish — CHANGES_REQUESTED** [review]

File: airflow/models/dag.py (hunk 2 above)
Comment (joshowen) [reply airflow/models/dag.py]: `existing_dags` is a dict in this code, so this is grabbing the first and only key which is a dag_id.  https://github.com/apache/airflow/pull/33242/files#diff-62c8e300ee91e0d59f81e0ea5d30834f04db71ae74f2e155a10b51056b00b59bR3053

**Review by dstandish — APPROVED** [review]

**Comment by dstandish** [comment]
hi @joshowen i moved the query builder to a function and added some testing for the query structure.  please take a look.  i changed the structure of the query slightly so that we don't have to join to the subquery in the 1 dag path (it can be just execution_date = (subquery))

**Comment by joshowen** [comment]
> hi @joshowen i moved the query builder to a function and added some testing for the query structure. please take a look. i changed the structure of the query slightly so that we don't have to join to the subquery in the 1 dag path (it can be just execution_date = (subquery))

lgtm.  Thanks!

**Pull request merged**